remote/blueprints/image.py

```diff
@@ -163,18 +209,53 @@ def resize_avatar():
     # Resize it to each size contained in `AvatarSize`.
 
     start = time.time()
+    avatars = {}
     try:
         # Confirmation basic avatar to resize.
         # Try to use max size avatar otherwise use original(rotated).
-        base_avatar = handler.auto_rotated().image
-        if _standard_avatar := _try_rescale(base_avatar, max(AvatarSize)):
-            base_avatar = ImageHandle.load_from_bytes(_standard_avatar)
-
-        avatars = {
-            f"avatar{size}": avatar_summary
-            for size in AvatarSize
-            if (avatar_summary := _get_avatar_summary(base_avatar, size))
-        }
+        if not handler.animated:
+            handler.auto_rotate()
+
+        image = handler.image
+        for size in AvatarSize:
+            if size.is_mandatory or _need_rescale(image, size):
+                # Choose rescale function based on format and animation.
+                if handler.animated and request.args.get("animated"):
+                    if handler.mime == ImageMIME.GIF or handler.mime == ImageMIME.WEBP:
+                        frames, durations = _rescale_animated_image_frames(
+                            image, size, resizeimage.resize_cover, validate=False
+                        )
+                    elif handler.mime == ImageMIME.PNG:
+                        frames, durations = _rescale_animated_png_frames(
+                            image, size, resizeimage.resize_cover, validate=False
+                        )
+                    else:
+                        frames = None
+
+                    if frames:
+                        data = _save_animated_frames_data(
+                            frames,
+                            durations,
+                            format=ImageMIME.PNG.pil_format,
+                            loop=0,  # always loop for avatar
+                        )
+                    else:
+                        # Fallback to first frame of animated image
+                        data = _rescale_single_frame_image(
+                            image,
+                            size,
+                            ImageMIME.PNG.pil_format,
+                            resizeimage.resize_cover,
+                            validate=False,
+                        )
+                else:
+                    data = _rescale_single_frame_image(
+                        image, size, resizeimage.resize_cover, validate=False
```

Comment:
Seems `_rescale_single_frame_image` function here signature don't match.

Reply:
`_rescale_single_frame_image` is the old `_try_rescale` that streamlines extracting (a) frame(s), processing (a) frame(s), and saving (a) frame(s). As there are various format specific ways to handle frames (buggy animated PNG) and different requirements on saving them (all avatars are converted to PNGs, but images just resize with original format unchanged), I have left it as is.

There is a better way to organize these methods as I described, but as we currently do not support _that_ many formats, I suggest we defer the refactor. New formats could bring surprise like PNG `default_mage` and break abstraction.

Reply:
Yes, I agree with you, but here the signature I said, I mean [_rescale_single_frame_image](https://github.com/v2ex/remote/pull/30/files#diff-68483a09b419f935fc7a94b6664ec28816286ad03f53bcef0933d6ab5a5144e3R291-R293) args signature,  the third position arg should be `format` not resized image, maybe like that:
```diff
data = _rescale_single_frame_image(
-                       image, size, resizeimage.resize_cover, validate=False
+                       image, size, handler.mime.pil_format, resizeimage.resize_cover, validate=False
)
```

Reply:
Thank you for pointing out!